eth_typing/abi.py

```diff
@@ -17,6 +17,39 @@
 """Binary data to be decoded."""
 
 
+class ABIEventComponent(TypedDict, total=False):
+    """
+    DEPRECATED: Use `ABIComponent`.
+
+    TypedDict to represent the `ABI` for nested event parameters.
+    Used as a component of `ABIEventParam`.
+    """
+
+    components: Sequence["ABIEventComponent"]
+    """List of nested event parameters for tuple event ABI types."""
+    name: str
+    """Name of the event parameter."""
+    type: str
+    """Type of the event parameter."""
+
+
+class ABIEventParam(TypedDict, total=False):
+    """
+    DEPRECATED: Use `ABIComponent`.
+
+    TypedDict to represent the `ABI` for event parameters.
+    """
+
+    indexed: bool
```

Comment:
We should try to capture `indexed` in the refactor to `ABIComponent`. Do you know if these objects always have `name`, `type` and `components`... and is it just a matter of whether they have values for them or not? I believe that's the case.

If so, we may want to drop `total=False` from `ABIComponent` and have an extended `ABIComponentIndexed` type that extends `ABIComponent` with `indexed` for the top-level components. This would require all components to have these fields and top level "input" and "output" components can use the `ABIComponentIndexed` class.

```py

class ABIComponent(TypedDict, total=True):
    name: str
    type: str
    components: Sequence["ABIComponent"]
    
class ABIComponentIndexed(ABIComponent):
    indexed: bool
```

Thoughts there?

Reply:
From the specs [where we where referencing](https://docs.soliditylang.org/en/latest/abi-spec.html#handling-tuple-types):

> The components of the tuple are then stored in the member `components`, which is of an array type and has the same structure as the top-level object except that `indexed` is not allowed there.

Reply:
Good catch, thank you for the suggestion. That captures exactly what we need.